File: Extensions/Tessa.Extensions.Default.Client/Workplaces/AutomaticNodeRefreshExtension.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.ComponentModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Threading;
using Tessa.Extensions.Default.Shared.Workplaces;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Views;
using Tessa.UI.Views.Extensions;
using Tessa.UI.Views.Workplaces.Tree;

namespace Tessa.Extensions.Default.Client.Workplaces
{
    /// <summary>
    /// Расширение, предоставляющее возможность автоматического обновления узлов рабочего места.
    /// </summary>
    /// <remarks>
    /// У расширения есть конфигуратор <see cref="AutomaticNodeRefreshExtensionConfigurator"/>.
    /// </remarks>
    public class AutomaticNodeRefreshExtension :
        ViewModel<EmptyModel>,
        ITreeItemExtension,
        IWorkplaceExtensionSettingsRestore
    {
        #region Fields

        private bool refreshPending;

        private AutomaticNodeRefreshSettings settings;

        private DispatcherTimer? timer;

        private ITreeItem? treeItem;

        private readonly ISessionFailedChecker sessionFailedChecker;

        #endregion

        #region Constructor

        /// <summary>
        /// Конструктор.
        /// </summary>
        /// <param name="sessionFailedChecker"><inheritdoc cref="ISessionFailedChecker" path="/summary"/></param>
        public AutomaticNodeRefreshExtension(ISessionFailedChecker sessionFailedChecker)
        {
            this.settings = new();
            this.sessionFailedChecker = NotNullOrThrow(sessionFailedChecker);
        }

        #endregion

        #region IWorkplaceExtension Implementation

        /// <inheritdoc />
        public void Clone(ITreeItem source, ITreeItem cloned, ICloneableContext context)
        {
        }

        /// <inheritdoc />
        public void Initialize(ITreeItem model) =>
            this.treeItem = model;

        /// <inheritdoc />
        public void Initialized(ITreeItem model)
        {
            this.timer = new DispatcherTimer(DispatcherPriority.ApplicationIdle)
            {
                Interval = TimeSpan.FromSeconds(this.settings.RefreshInterval)
            };

            this.timer.Tick += async (s, e) => await this.UpdateByTimerAsync(false);

            this.SubscribeToEvents(model);
            if (model?.Workplace.IsActive is true)
            {
                // вкладка с рабочим местом активна на момент запуска приложения
                this.StartTimer();
            }
        }

        #endregion

        #region IWorkplaceExtensionSettingsRestore Implementation

        /// <inheritdoc />
        public void Restore(Dictionary<string, object?> metadata) =>
            this.settings = metadata.FromSerializedDictionary<AutomaticNodeRefreshSettings>();

        #endregion

        #region Base Overrides

        /// <inheritdoc />
        protected override bool OnReceiveWeakEvent(Type managerType, object sender, EventArgs e)
        {
            if (managerType != typeof(PropertyChangedEventManager))
            {
                return base.OnReceiveWeakEvent(managerType, sender, e);
            }

            var eventArgs = (PropertyChangedEventArgs) e;
            switch (eventArgs.PropertyName)
            {
                case "Parent":
                    var treeItem = (ITreeItem) sender;
                    if (treeItem.Parent is null)
                    {
                        this.UnSubscribeFromEvents(treeItem);
                        this.StopTimer();
                    }
                    else
                    {
                        this.SubscribeToEvents(treeItem);
                        this.StartTimer();
                    }

                    break;
                case "IsActive":
                    var workplace = (IWorkplaceViewModel) sender;
                    if (workplace.IsActive)
                    {
                        this.StartAndUpdateIfRequired(true);
                    }

                    break;
                case "IsExpanded":
                    var isExpanded = ((ITreeItem) sender).IsExpanded;
                    if (isExpanded && this.treeItem?.IsVisibleInPath() is true)
                    {
                        this.StartAndUpdateIfRequired(false);
                    }

                    break;
                case "LastUpdateTime":
                    if (this.timer?.IsEnabled is true)
                    {
                        this.StopTimer();
                        this.StartTimer();
                        this.refreshPending = false;
                    }

                    break;
                case "IsSelected":
                    var isSelected = ((ITreeItem) sender).IsSelected;
                    if (isSelected && this.timer?.IsEnabled is not true)
                    {
                        this.StartAndUpdateIfRequired(false);
                    }

                    break;
            }

            return true;
        }

        #endregion

        #region Private Methods

        /// <summary>
        /// Осуществляет подписку на события изменения рабочего места и активности узла.
        /// </summary>
        /// <param name="treeItem">Узел дерева рабочего места.</param>
        /// <exception cref="ArgumentNullException"/>
        private void SubscribeToEvents(ITreeItem treeItem)
        {
            ThrowIfNull(treeItem);

            PropertyChangedEventManager.AddListener(treeItem, this, "LastUpdateTime");
            PropertyChangedEventManager.AddListener(treeItem, this, "IsExpanded");
            PropertyChangedEventManager.AddListener(treeItem, this, "IsSelected");
            PropertyChangedEventManager.AddListener(treeItem, this, "Parent");
            if (treeItem.Workplace is not null)
            {
                PropertyChangedEventManager.AddListener(treeItem.Workplace, this, "IsActive");
            }

            var currentNode = treeItem.Parent;
            while (currentNode is not null)
            {
                PropertyChangedEventManager.AddListener(currentNode, this, "IsExpanded");
                currentNode = currentNode.Parent;
            }

            this.SubscribeToChildEvents(treeItem);
        }

        private void SubscribeToChildEvents(ITreeItem treeItem)
        {
            this.SubscribeToChildChangingEvents(treeItem, true);

            foreach (var child in treeItem.Items)
            {
                this.SubscribeToChildChangingEvents(child, false);

                PropertyChangedEventManager.AddListener(child, this, "IsExpanded");
                PropertyChangedEventManager.AddListener(child, this, "IsSelected");

                if (child is ISubsetTreeItem subsetTreeItem)
                {
                    foreach (var subsetChild in subsetTreeItem.Items)
                    {
                        PropertyChangedEventManager.AddListener(subsetChild, this, "IsSelected");
                    }
                }
            }
        }

        private void SubscribeToChildChangingEvents(ITreeItem treeItem, bool firstLevel)
        {
            ((INotifyCollectionChanged) treeItem.Items).CollectionChanged += (_, e) =>
            {
                if (e.NewItems is { Count: > 0 } newItems)
                {
                    var isExpanded = false;
                    foreach (var item in newItems.Cast<ITreeItem>())
                    {
                        PropertyChangedEventManager.AddListener(item, this, "IsSelected");

                        if (firstLevel)
                        {
                            PropertyChangedEventManager.AddListener(item, this, "IsExpanded");

                            if (item is ISubsetTreeItem subsetTreeItem)
                            {
                                this.SubscribeToChildChangingEvents(subsetTreeItem, false);
                            }
                        }

                        isExpanded |= item.IsExpanded;
                    }

                    if (firstLevel && isExpanded)
                    {
                        this.StartAndUpdateIfRequired(false);
                    }
                }

                if (e.OldItems is { Count: > 0 } oldItems)
                {
                    foreach (var item in oldItems.Cast<ITreeItem>())
                    {
                        PropertyChangedEventManager.RemoveListener(item, this, "IsSelected");

                        if (firstLevel)
                        {
                            PropertyChangedEventManager.RemoveListener(item, this, "IsExpanded");

                            if (item is ISubsetTreeItem subsetTreeItem)
                            {
                                foreach (var subsetChild in subsetTreeItem.Items)
                                {
                                    PropertyChangedEventManager.RemoveListener(subsetChild, this, "IsSelected");
                                }
                            }
                        }
                    }
                }
            };
        }

        /// <summary>
        /// Отписывается от событий влияющих на запуск автоматического обновления.
        /// </summary>
        /// <param name="treeItem">Узел дерева.</param>
        /// <exception cref="ArgumentNullException"/>
        private void UnSubscribeFromEvents(ITreeItem treeItem)
        {
            ThrowIfNull(treeItem);

            PropertyChangedEventManager.RemoveListener(treeItem, this, "Parent");
            PropertyChangedEventManager.RemoveListener(treeItem, this, "LastUpdateTime");
            PropertyChangedEventManager.RemoveListener(treeItem, this, "IsSelected");
            if (treeItem.Workplace is not null)
            {
                PropertyChangedEventManager.RemoveListener(treeItem.Workplace, this, "IsActive");
            }

            var currentNode = treeItem.Parent;
            while (currentNode is not null)
            {
                PropertyChangedEventManager.RemoveListener(currentNode, this, "IsExpanded");
                currentNode = currentNode.Parent;
            }

            foreach (var child in treeItem.Items)
            {
                PropertyChangedEventManager.RemoveListener(child, this, "IsExpanded");
                PropertyChangedEventManager.RemoveListener(child, this, "IsSelected");

                if (child is ISubsetTreeItem subsetTreeItem)
                {
                    foreach (var subsetChild in subsetTreeItem.Items)
                    {
                        PropertyChangedEventManager.RemoveListener(subsetChild, this, "IsSelected");
                    }
                }
            }
        }

        /// <summary>
        /// Вызывается при срабатывании таймера.
        /// </summary>
        /// <param name="skipUpdateTable">
        /// Признак необходимости отменить обновление таблицы.
        /// </param>
        private async Task UpdateByTimerAsync(bool skipUpdateTable)
        {
            // Если задача не успела отработать или узел находится в процессе обновления, то просто выходим из задачи.
            // При сравнении использован коэффициент 0.98 (время уменьшено на 1/50), потому что обновление
            // treeItem.LastUpdateTime происходит через некоторое время после того, как предыдущее расширение отработало,
            // поэтому при проверке нужен доверительный интервал.
            if (this.treeItem is null ||
                this.treeItem.InUpdate ||
                (DateTime.UtcNow - this.treeItem.LastUpdateTime).TotalSeconds < this.settings.RefreshInterval * 0.98 ||
                await this.sessionFailedChecker.IsCurrentSessionFailedAsync())
            {
                return;
            }

            // Останавливаемся, если узел РМ не активно или узел не виден в дереве.
            if (!this.treeItem.Workplace.IsActive
                || !this.treeItem.IsVisibleInPath())
            {
                this.refreshPending = true;
                this.StopTimer();
                return;
            }

            // Останавливаемся, если не включена настройка "Обновлять всегда" и не выбран элемент с указанным расширением и он не раскрыт вместе с сабсетами.
            if (!this.settings.AlwaysRefresh
                && !this.treeItem.HasSelection()
                && (!this.treeItem.IsExpanded || !this.treeItem.Items.Any(i => i.IsExpanded)))
            {
                this.refreshPending = true;
                this.StopTimer();
                return;
            }

            await using var _ = SessionRequestTypeContext.Create(SessionRequestType.Background);

            await this.treeItem.RefreshNodeAsync(onCompletedAsync: async t =>
            {
                if (this.treeItem.HasSelection())
                {
                    await this.RefreshTableContentAsync(skipUpdateTable);
                }
            });
        }

        /// <summary>
        /// Запускает таймер.
        /// </summary>
        private void StartTimer()
        {
            if (this.timer is not null && !this.timer.IsEnabled)
            {
                this.timer.Start();
            }
        }

        /// <summary>
        /// Останавливает таймер.
        /// </summary>
        private void StopTimer() =>
            this.timer?.Stop();

        /// <summary>
        /// Обновляет контент в таблице.
        /// </summary>
        /// <param name="skipUpdateTable">Пропустить обновление.</param>
        private Task RefreshTableContentAsync(bool skipUpdateTable)
        {
            this.refreshPending = false;

            return this.settings.WithContentDataRefreshing && !skipUpdateTable
                ? RefreshContentAsync(this.treeItem?.Workplace)
                : Task.CompletedTask;
        }

        /// <summary>
        /// Вызывает обновление содержимого табличной части.
        /// </summary>
        /// <param name="workplaceViewModel">Модель рабочего места.</param>
        private static async Task RefreshContentAsync(IWorkplaceViewModel? workplaceViewModel)
        {
            if (workplaceViewModel is null
                || workplaceViewModel.Context.ViewContext is not { } viewContext)
            {
                return;
            }

            // Получаем верхнюю вью (от которой зависят остальные)
            var rootContext = viewContext.GetRoot();
            var viewComponent = rootContext.TryGetViewContainer();

            if (viewComponent is null || viewComponent.CurrentPage == 1)
            {
                // либо вью не поддерживает пейджинг, либо страница и так первая, либо это какой-то кастом
                // если кастом, то надеемся, что он поддерживает RefreshCommand
                await rootContext.RefreshViewAsync();
            }
            else
            {
                // Refresh будет автоматом при изменении номера страницы на первую
                viewComponent.CurrentPage = 1;
            }
        }

        /// <summary>
        /// Вызывает обновление и стартует таймер в случае необходимости.
        /// </summary>
        /// <param name="skipUpdateTable">Пропустить обновление табличной части.</param>
        private void StartAndUpdateIfRequired(bool skipUpdateTable)
        {
            if (this.refreshPending)
            {
                //this.refreshPending = false;
                var _ = this.UpdateByTimerAsync(skipUpdateTable);
            }

            this.StartTimer();
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workplaces/automaticNodeRefreshExtension.ts
import { reaction } from 'mobx';
import { TreeItemExtension } from 'tessa/ui/views/extensions';
import {
  ISubsetTreeItem,
  ITreeItem,
  isTreeItemVisibleInPath
} from 'tessa/ui/views/workplaces/tree';
import { IStorage } from 'tessa/platform/storage';
import { tryGetFromSettings } from 'tessa/ui';
import { IWorkplaceViewModel, IWorkplaceViewComponent } from 'tessa/ui/views';
import {
  extension,
  HttpHooksContext,
  HttpRequestHeaders,
  inject,
  ISessionHealth,
  ISessionHealth$
} from '@tessa/application';

@extension()
export class AutomaticNodeRefreshExtension extends TreeItemExtension {
  private _settings: AutomaticNodeRefreshSettings;

  private _timer: any | null;

  private _refreshPending = false;

  private _treeItem: ITreeItem;

  private _disposes: Function[] = [];

  //#region ctor

  constructor(@inject(ISessionHealth$) private readonly _sessionHealth: ISessionHealth) {
    super();
  }

  //#endregion

  public getExtensionName(): string {
    return 'Tessa.Extensions.Default.Client.Workplaces.AutomaticNodeRefreshExtension';
  }

  public initialized(model: ITreeItem): void {
    this._treeItem = model;
    this._settings = new AutomaticNodeRefreshSettings(this.settingsStorage);
    this.subscribeToEvents(model);
    reaction(
      () => model.parent,
      parent => {
        this.unsubscribeFromEvents();
        if (!parent) {
          this.stopTimer();
        } else {
          this.subscribeToEvents(model);
          this.startTimer();
        }
      }
    );
    if (model.workplace && model.workplace.isActive) {
      // вкладка с рабочим местом активна на момент запуска приложения
      this.startTimer();
    }
  }

  private addIsExpandedListener(treeItem: ITreeItem) {
    this._disposes.push(
      reaction(
        () => treeItem.isExpanded,
        isExpanded => {
          if (isExpanded && isTreeItemVisibleInPath(treeItem)) {
            if (this._refreshPending) {
              this.updateByTimer(false);
            }
            this.startTimer();
          }
        }
      )
    );
  }

  private addIsSelectedListener(treeItem: ITreeItem) {
    this._disposes.push(
      reaction(
        () => treeItem.isSelected,
        isSelected => {
          if (isSelected) {
            if (this._refreshPending) {
              this.updateByTimer(false);
            }
            this.startTimer();
          }
        }
      )
    );
  }

  private subscribeToEvents(treeItem: ITreeItem) {
    this._disposes.push(
      reaction(
        () => treeItem.workplace.isActive,
        isActive => {
          if (isActive) {
            if (this._refreshPending) {
              this.updateByTimer(true);
            }
            this.startTimer();
          }
        }
      )
    );

    this._disposes.push(
      reaction(
        () => treeItem.isLoading,
        isLoading => {
          if (!isLoading && !this._timer) {
            this.stopTimer();
            this.startTimer();
            this._refreshPending = false;
          }
        }
      )
    );

    let currentNode = treeItem.parent;
    while (currentNode) {
      this.addIsExpandedListener(currentNode);
      currentNode = currentNode.parent;
    }

    this.addIsExpandedListener(treeItem);
    this.addIsSelectedListener(treeItem);

    this.subscribeToChildEvents(treeItem);
  }

  private subscribeToChildEvents(treeItem: ITreeItem) {
    this.subscribeToChildChangingEvents(treeItem, true);

    for (const child of treeItem.items) {
      this.subscribeToChildChangingEvents(child, false);

      this.addIsExpandedListener(child);
      this.addIsSelectedListener(child);

      const subsetTreeItem = child as ISubsetTreeItem;
      if (subsetTreeItem) {
        for (const subsetChild of subsetTreeItem.items) {
          this.addIsSelectedListener(subsetChild);
        }
      }
    }
  }

  private subscribeToChildChangingEvents(treeItem: ITreeItem, firstLevel: boolean) {
    this._disposes.push(
      reaction(
        () => treeItem.visibleItems,
        items => {
          let isExpanded = false;
          for (const child of items) {
            this.addIsSelectedListener(child);
            if (firstLevel) {
              this.addIsExpandedListener(child);

              const subsetTreeItem = child as ISubsetTreeItem;
              if (subsetTreeItem) {
                this.subscribeToChildChangingEvents(subsetTreeItem, false);
              }
            }

            isExpanded = isExpanded || child.isExpanded;
          }
          if (firstLevel && isExpanded) {
            if (this._refreshPending) {
              this.updateByTimer(false);
            }
            this.startTimer();
          }
        }
      )
    );
  }

  private unsubscribeFromEvents() {
    for (const dispose of this._disposes) {
      dispose();
    }
    this._disposes.length = 0;
  }

  private startTimer() {
    if (!this._timer) {
      this._timer = setTimeout(
        () => this.updateByTimer(false),
        this._settings.refreshInterval * 1000
      );
    }
  }

  private stopTimer() {
    if (this._timer) {
      clearTimeout(this._timer);
      this._timer = null;
    }
  }

  private async updateByTimer(skipUpdateTable: boolean) {
    this.stopTimer();
    if (await this.updateByTimerCore(skipUpdateTable)) {
      this.startTimer();
    }
  }

  private async updateByTimerCore(skipUpdateTable: boolean): Promise<boolean> {
    // Если задача не успела отработать или узел находится в процессе обновления, то просто выходим из задачи.
    // При сравнении использован коэффициент 0.98 (время уменьшено на 1/50), потому что обновление
    // treeItem.LastUpdateTime происходит через некоторое время после того, как предыдущее расширение отработало,
    // поэтому при проверке нужен доверительный интервал.
    if (
      this._treeItem.isLoading ||
      Date.now() - this._treeItem.lastUpdateTime < this._settings.refreshInterval * 1000 * 0.98 ||
      !this._sessionHealth.isHealthy
    ) {
      return true;
    }

    // Останавливаемся, если узел РМ не активно или узел не виден в дереве.
    if (!this._treeItem.workplace.isActive || !isTreeItemVisibleInPath(this._treeItem)) {
      this._refreshPending = true;
      return false;
    }

    // Останавливаемся, если не включена настройка "Обновлять всегда" и не выбран элемент с указанным расширением и он не раскрыт вместе с сабсетами.
    if (
      !this._settings.alwaysRefresh &&
      !this._treeItem.hasSelection() &&
      (!this._treeItem.isExpanded || !this._treeItem.items.some(i => i.isExpanded))
    ) {
      this._refreshPending = true;
      return false;
    }

    const scope = HttpHooksContext.create(
      new HttpHooksContext([
        {
          beforeRequest: async ctx =>
            ctx.request.headers.append(HttpRequestHeaders.RequestType, 'background')
        }
      ])
    );
    scope.run(async () => {
      await this._treeItem.refreshNode();
      if (this._treeItem.hasSelection()) {
        this.refreshTableContent(skipUpdateTable);
      }
    });

    return true;
  }

  private async refreshTableContent(skipUpdateTable: boolean): Promise<void> {
    this._refreshPending = false;
    if (this._settings.withContentDataRefreshing && !skipUpdateTable) {
      await this.refreshContent(this._treeItem.workplace);
    }
  }

  private async refreshContent(workplaceViewModel: IWorkplaceViewModel): Promise<void> {
    if (!workplaceViewModel) {
      return;
    }

    // обновляем содержимое (таблицы)
    const viewContext = workplaceViewModel.context.viewContext;
    if (viewContext) {
      // получаем верхнюю вью (от которой зависят остальные)
      let rootContext = viewContext;
      while (rootContext.parentContext) {
        rootContext = rootContext.parentContext;
      }
      const viewComponent = rootContext as IWorkplaceViewComponent;

      if (viewComponent.currentPage == undefined || viewComponent.currentPage === 1) {
        // либо вью не поддерживает пейджинг, либо страница и так первая, либо это какой-то кастом
        // если кастом, то надеемся, что он поддерживает RefreshCommand
        await rootContext.refreshView();
      } else {
        // Refresh будет автоматом при изменении номера страницы на первую
        viewComponent.currentPage = 1;
      }
    }
  }
}

class AutomaticNodeRefreshSettings {
  constructor(storage: IStorage) {
    this.refreshInterval = tryGetFromSettings(storage, 'RefreshInterval', 300);
    this.withContentDataRefreshing = tryGetFromSettings(storage, 'WithContentDataRefreshing', true);
    this.alwaysRefresh = tryGetFromSettings(storage, 'AlwaysRefresh', false);
  }

  public refreshInterval: number;

  public withContentDataRefreshing: boolean;

  public alwaysRefresh: boolean;
}
